[Refactor] Added support for OpenAI ChatCompletions

DaveOkpare opened the pull request:
This PR adds support for OpenAI ChatCompletion requested by @baberabb in #737

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=839) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=839) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=839) it.</sub>

DaveOkpare commented:
> This looks great, wondering if you can test this on any task of your choice?

@lintangsutawika Apologies for the late reply, I've fixed my code and but kept running out-of-storage when downloading the dataset for `greedy_until`. Can you please point me to a guide to run a test on low-resource?

DaveOkpare commented:
> This is great! Thanks for this! It failed to run though as it checks for duplicated aliases and you are re-using `openai` and `gooseai`.
> 
> After removing those I'm getting: `openai.error.InvalidRequestError: '$.prompt' is invalid. Please check the API reference: https://platform.openai.com/docs/api-reference.` Also seems there is no way to pass the `async` arg to the model?
> 
> You can always test it out with `--limit` to save API calls.

Thanks for the help, I've fixed the error.  I've removed the `is_async` arg, it was just a flag to reuse the same function but with async.

StellaAthena commented:
@DaveOkpare So have you run this end-to-end and confirmed it worked?
